syncer/streamer_controller.go

```diff
@@ -0,0 +1,339 @@
+	}()
+
+	// FIXME: can enable GTID
+	if r.EnableGTID {
+		// NOTE: our (per-table based) checkpoint does not support GTID yet
+		return nil, terror.ErrSyncerUnitRemoteSteamerWithGTID.Generate()
+	}
+
+	// position's name may contain uuid, so need remove it
+	adjustedPos := binlog.AdjustPosition(pos)
+	streamer, err := r.reader.StartSync(adjustedPos)
+	return streamer, terror.ErrSyncerUnitRemoteSteamerStartSync.Delegate(err)
+}
+
+// StreamerController controls the streamer for read binlog, include:
+// 1. reopen streamer
+// 2. redirect binlog position or gtid
+// 3. transfor from local streamer to remote streamer
+type StreamerController struct {
+	sync.RWMutex
+
+	// the initital binlog type
+	initBinlogType BinlogType
+
+	// the current binlog type
+	currentBinlogType BinlogType
+
+	syncCfg        replication.BinlogSyncerConfig
+	localBinlogDir string
+	timezone       *time.Location
+
+	streamer         streamer.Streamer
+	streamerProducer StreamerProducer
+
+	// meetError means meeting error when get binlog event
+	// if binlogType is local and meetError is true, then need to create remote binlog stream
+	meetError bool
+
+	fromDB *UpStreamConn
+
+	uuidSuffix string
+}
+
+// NewStreamerController creates a new streamer controller
+func NewStreamerController(tctx tcontext.Context, syncCfg replication.BinlogSyncerConfig, fromDB *UpStreamConn, binlogType BinlogType, localBinlogDir string, timezone *time.Location, beginPos mysql.Position) (*StreamerController, error) {
+	streamerController := &StreamerController{
+		initBinlogType:    binlogType,
+		currentBinlogType: binlogType,
+		syncCfg:           syncCfg,
+		localBinlogDir:    localBinlogDir,
+		timezone:          timezone,
+		fromDB:            fromDB,
+	}
+
+	err := streamerController.ResetReplicationSyncer(tctx, beginPos)
+	if err != nil {
+		streamerController.Close(tctx)
+		return nil, err
+	}
+
+	return streamerController, nil
+}
+
+// ResetReplicationSyncer reset the replication
+func (c *StreamerController) ResetReplicationSyncer(tctx tcontext.Context, pos mysql.Position) (err error) {
+	c.Lock()
+	defer c.Unlock()
+
+	uuidSameWithUpstream := true
+
+	// close old streamerProducer
+	if c.streamerProducer != nil {
+		switch t := c.streamerProducer.(type) {
+		case *remoteBinlogReader:
+			c.closeBinlogSyncer(tctx, t.reader)
+		case *localBinlogReader:
+			// check the uuid before close
+			uuidSameWithUpstream, err = c.checkUUIDSameWithUpstream(pos, t.reader.GetUUIDs())
+			if err != nil {
+				return err
+			}
+			// TODO: close old local reader before creating a new one
+		default:
+			// some other producers such as mockStreamerProducer, should not re-create
+			c.streamer, err = c.streamerProducer.generateStreamer(pos)
+			return err
+		}
+	}
+
+	// binlog type is local: means relay already use the server id, so need change to a new server id
+	// binlog type is remote: dm-worker has more than one sub task, so need generate random server id
```

Comment:
In fact:
- in local mode, this `server-id` is never used.
- in remote mode, the subtask can reuse the one it used before and only need to generate a new one for different subtasks.

so, should we need to avoid generating too many `server-id` (and add too many slaves for the upstream server) for some operations (like `RedirectStreamer`)?

Reply:
Is it still generating random `server-id` when redirecting now?

Reply:
I refine it in 439b166, only generate random server id once in `Init`

Reply:
refine in 394bb01, only generate random server id when create streamer controller